### ISSUE TITLE
ps aux hangs when running docker container with resource limit

### ISSUE THREAD

**Issue by skilledno1** [moby/moby#34579]
**Description**
Hi,  I run a java application container  with resource limit,  as lots of  oom-kill events happen every 2 or 3 minutes,   then execute ps aux, the command hangs.  strace is 

```
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=388, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=388, ...}) = 0
write(1, "root     43119  0.0  0.0 356584 "..., 177root     43119  0.0  0.0 356584  2420 ?        Sl   8月18   0:00 docker-containerd-shim 433759757a92eea05d18e978dae15a4820a3c5f24f6145317ecff119500947b2 /var/run/docker/libcon
) = 177
stat("/proc/43136", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
open("/proc/43136/stat", O_RDONLY)      = 6
read(6, "43136 (pause) S 43119 43136 4313"..., 2048) = 313
close(6)                                = 0
open("/proc/43136/status", O_RDONLY)    = 6
read(6, "Name:\tpause\nState:\tS (sleeping)\n"..., 2048) = 1065
close(6)                                = 0
open("/proc/43136/cmdline", O_RDONLY)   = 6
read(6, "/pause\0", 131072)             = 7
read(6, "", 131065)                     = 0
close(6)                                = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=388, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=388, ...}) = 0
write(1, "root     43136  0.0  0.0   1020 "..., 73root     43136  0.0  0.0   1020     4 ?        Ss   8月18   0:00 /pause
) = 73
stat("/proc/43140", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
open("/proc/43140/stat", O_RDONLY)      = 6
read(6, "43140 (exe) D 1 43140 43140 0 -1"..., 2048) = 319
close(6)                                = 0
open("/proc/43140/status", O_RDONLY)    = 6
read(6, "Name:\texe\nState:\tD (disk sleep)\n"..., 2048) = 1060
close(6)                                = 0
open("/proc/43140/cmdline", O_RDONLY)   = 6
read(6, 
```
43140 process state is D (disk sleep), and stack is 

```
[root@k8s-dbg-master-1 ~]# cat /proc/43140/stack 
[<ffffffff811688be>] sleep_on_page+0xe/0x20
[<ffffffff81168646>] wait_on_page_bit+0x86/0xb0
[<ffffffff8117d472>] shrink_page_list+0x6c2/0xad0
[<ffffffff8117df3a>] shrink_inactive_list+0x1ea/0x560
[<ffffffff8117ea05>] shrink_lruvec+0x375/0x760
[<ffffffff8117ee66>] shrink_zone+0x76/0x1a0
[<ffffffff8117f370>] do_try_to_free_pages+0xf0/0x4e0
[<ffffffff8117f9aa>] try_to_free_mem_cgroup_pages+0xca/0x160
[<ffffffff811d207e>] mem_cgroup_reclaim+0x4e/0xe0
[<ffffffff811d253c>] __mem_cgroup_try_charge+0x42c/0x650
[<ffffffff811d2e89>] mem_cgroup_charge_common+0x59/0xc0
[<ffffffff811d4c56>] mem_cgroup_newpage_charge+0x26/0x30
[<ffffffff81197676>] handle_mm_fault+0xba6/0xf50
[<ffffffff81640e22>] __do_page_fault+0x152/0x420
[<ffffffff81641113>] do_page_fault+0x23/0x80
[<ffffffff8163d408>] page_fault+0x28/0x30
[<ffffffffffffffff>] 0xffffffffffffffff
```
the container process state is S (sleeping)，and stack is 

```
[root@k8s-dbg-master-1 ~]# cat /proc/19879/stack
[<ffffffff810e2464>] futex_wait_queue_me+0xc4/0x120
[<ffffffff810e2fd9>] futex_wait+0x179/0x280
[<ffffffff810e506e>] do_futex+0xfe/0x5b0
[<ffffffff810e55a0>] SyS_futex+0x80/0x180
[<ffffffff81645909>] system_call_fastpath+0x16/0x1b
[<ffffffffffffffff>] 0xffffffffffffffff
```

key message:

```
8月 21 15:27:09 k8s-dbg-master-1 docker[1115]: time="2017-08-21T15:27:09.632857520+08:00" level=error msg="Handler for POST /v1.24/exec/36e1fae72a29a0517d3af36cced2362ce9587d0299ff882df6c257bbc2b00e7a/resize returned error: rpc error: code = 2 desc = containerd: process not found for container"
```

oom messages:

```
[root@k8s-dbg-master-1 ~]# grep oom-kill /var/log/messages
Aug 21 10:43:29 k8s-dbg-master-1 kernel: java invoked oom-killer: gfp_mask=0xd0, order=0, oom_score_adj=999
Aug 21 10:47:53 k8s-dbg-master-1 kernel: java invoked oom-killer: gfp_mask=0xd0, order=0, oom_score_adj=999
.......
Aug 21 14:48:46 k8s-dbg-master-1 kernel: java invoked oom-killer: gfp_mask=0xd0, order=0, oom_score_adj=999  ----  the last oom-kill event
```

**Steps to reproduce the issue:**

the problem is easy to reproduce with my own java application, if need any more info， please let me know。

**Output of `docker version`:**

```
[root@k8s-dbg-master-1 ~]# docker -D version
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:20:01 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:20:01 2017
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 60
 Running: 42
 Paused: 0
 Stopped: 18
Images: 201
Server Version: 1.12.6
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 25.06 GB
 Data Space Total: 892.3 GB
 Data Space Available: 867.3 GB
 Metadata Space Used: 9.986 MB
 Metadata Space Total: 9.391 GB
 Metadata Space Available: 9.381 GB
 Thin Pool Minimum Free Space: 89.23 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 31.21 GiB
Name: k8s-dbg-master-1
ID: 5WDW:F3AZ:3THZ:SFVR:LZTC:X3BW:OQJD:XFLH:ELPF:L2CK:YE6N:GUWD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 10.233.0.0/18
 172.0.0.0/8
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
```
physical machine with system info:
[root@k8s-master-1 ~]# uname -a
Linux k8s-master-1 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
[root@k8s-master-1 ~]# cat /etc/redhat-release 
CentOS Linux release 7.2.1511 (Core)
```

**Comment by skilledno1** [comment]
@AkihiroSuda  the issue info was updated， please reopen this issue，if need any other info，please let me know， thx。

**Comment by skilledno1** [comment]
@AkihiroSuda  i have close khugepage use "echo never > /sys/kernel/mm/redhat_transparent_hugepage/enabled", but it does not affect, problem still exist,

**Comment by franciscogimeno** [comment]
@skilledno1 , what's the main task (reserving memory, IOPS...) the application is doing when it gets OOMed?
how much memory is it consuming?
what's the memory limit to the container?
what's the memory usage in the node?

**Comment by miroswan** [comment]
Seeing a similar issue. What is are the contents of the /proc/$PID/syscall file when this happens?

**Comment by marenzo** [comment]
Same here with newer Docker version:

`Docker version 18.06.0-ce, build 0ffa825`
`3.13.0-149-generic #199-Ubuntu SMP Thu May 17 10:12:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
`Description:	Ubuntu 14.04.5 LTS`

Does someone knows what may be the root cause?

**Comment by Nowaker** [comment]
I've seen Docker deadlocked hundreds of time over the last half a year.

When this happens `ps aux` deadlocks. `ps aux -q $DOCKERSPID` deadlocks. `ps -A q $DOCKERSPID` works because this does not read `/proc/$DOCKERSPID/cmdline`.

> Seeing a similar issue. What is are the contents of the /proc/$PID/syscall file when this happens?

Don't know about that one, but we do have `/proc/$PID/stack`:

```
[<0>] futex_wait_queue_me+0xc4/0x120
[<0>] futex_wait+0x10a/0x250
[<0>] do_futex+0x325/0x500
[<0>] SyS_futex+0x13b/0x180
[<0>] do_syscall_64+0x73/0x130
[<0>] entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[<0>] 0xffffffffffffffff
```

Futex wait on God knows what. Unfortunately, we don't have a `strace` output.

When this happens, Docker daemon will stay deadlocked for half an hour up to an hour. The only way to recover quickly is `kill -9`.

**Comment by mateuszkwiatkowski** [comment]
Some strace output when docker is locked (`strace -f -p <docker pid>`):
```
[pid  2743] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2742] futex(0xc4208ca548, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2728] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2727] futex(0xc4208ca148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2726] futex(0xc4208b6148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2725] futex(0xc4208a4148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2724] futex(0xc42088c148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2723] futex(0xc420498548, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2722] futex(0xc4204a6548, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2721] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2720] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2719] futex(0xc420484948, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2718] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2717] futex(0xc420128d48, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2716] futex(0xc4204ca548, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2715] futex(0xc4204d8d48, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2712] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2710] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2709] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2708] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2707] futex(0xc4204ca148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2706] futex(0xc4204bc148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2705] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2704] futex(0xc420498148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2703] futex(0xc420484148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2697] futex(0x55f2627a76b8, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2696] futex(0xc420128148, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2695] futex(0xc4200eed48, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  2693] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2691] futex(0x55f26278edc0, FUTEX_WAIT, 2, NULL <unfinished ...>
[pid  2654] futex(0x55f262786f08, FUTEX_WAIT, 0, NULL
```
Note: pid 2654, after which strace stopps is dockerd itself. The other pids don't exist anymore.

syscall file for dockerd process:
```
 cat /proc/2654/syscall
202 0x55f262786f08 0x0 0x0 0x0 0x0 0x0 0x7ffe8863ed38 0x55f25f9ab2c3
```

No entries in dmesg during the lockup. Last entry in dockerd log:
```
Jun 21 03:08:01 node-2.prod-bm.kubecluster.vk.dreamhost.io dockerd[2654]: time="2019-06-21T03:08:01.025041687-07:00" level=warning msg="got error while decoding json" error="unexpected EOF" retries=0
```

Dockerd process is in sleeping state:
root@node-2:/proc/2654# cat status
Name:   dockerd
Umask:  0022
State:  S (sleeping)

**Comment by rawmind** [comment]
I caught the deadlock.
It's reproducible in 99% cases

I've the same behavior mentioned by @mateuszkwiatkowski

```
[root@jenkins meetingtown-master]# ps ax | grep -e "docker\|containerd"
 3849 ?        Ssl    0:00 /usr/bin/containerd
 3854 ?        Ssl    0:02 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock
 9271 pts/2    Tl     0:00 containerd
 9282 pts/2    S+     0:00 grep --color=auto -e docker\|containerd
14932 pts/0    S+     0:00 strace docker build -t some:tst -f docker/Dockerfile .
14934 pts/0    Sl+    0:00 docker build -t some:tst -f docker/Dockerfile .
14951 ?        Sl     0:00 containerd-shim -namespace moby -workdir /var/lib/containerd/io.containerd.runtime.v1.linux/moby/d72f9d31e6ecb6c6984e779ff39866071c89e8dd5235194d602ec64bb3c93418 -address /run/containerd/containerd.sock -containerd-binary /usr/bin/containerd -runtime-root /var/run/docker/runtime-runc
```

strace 
```
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x5631e845be88, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
```

[root@jenkins meetingtown-master]# cat /proc/14934/stack 
```
[<ffffffffb7712086>] futex_wait_queue_me+0xc6/0x130
[<ffffffffb7712e2b>] futex_wait+0x17b/0x280
[<ffffffffb7714b76>] do_futex+0x106/0x5a0
[<ffffffffb7715090>] SyS_futex+0x80/0x190
[<ffffffffb7d8e172>] tracesys+0xa6/0xcc
[<ffffffffffffffff>] 0xffffffffffffffff
```

any suggestions how to spot the cause?
